r/R/dplyr-summarize.R

```diff
@@ -136,6 +136,17 @@ summarize_eval <- function(name, quosure, ctx, recurse = FALSE) {
     return()
   }
 
+  if ("median" %in% funs_in_expr) {
+    expr <- wrap_median(expr)
```

Comment:
@nealrichardson this implementation will not actually work until #11159 is merged and we have a `list_element` scalar kernel, but the gist of this as follows:

We need to translate `median(x, na.rm)` to `list_element(tdigest(x, q = 0.5, skip_nulls = na.rm), 0)` but because of how aggregate function bindings work, it isn't possible to do that directly in the definition for `agg_funcs$median`, so I hacked together this solution in which `summarize_eval()` calls a function named `wrap_median()` that recurses through the expression AST and replaces `median(...)` with `arrow_list_element(median(...))`.

Can you think of a more straightforward immediate solution to this?

I think we should ask for a `median()` kernel to avoid having to do all these contortions in the R bindings.

Reply:
My plan for this is to special case quantile and median in `extract_aggregations`, and instead of inserting `expr <- as.symbol(tmpname)`, i.e. just the temp variable name, we would insert `str2lang(paste0("arrow_list_element(", tmpname, ", 0L)"))`

Reply:
Yep, my first instinct here was to use `extract_aggregations()` to do this (since you're already recursing through the AST there), but that is only called called when `any(inner_agg)`. I'll change it so it's also called in the median case, and I'll do the AST insertion in there.

Reply:
@nealrichardson the code I needed to add inside `wrap_median()` and in various other places here to deal with the `tdigest` and `hash_tdigest` kernel output  is unpleasant stuff.  I think it's best to keep `wrap_median()` self-contained instead of contaminating `extract_aggregations()` with it.

I think we should ask for pair of `appx_median` and `hash_appx_median` kernels that wrap `tdigest(q = 0.5)` and `hash_tdigest(q = 0.5)`. That would allow us to clean up this mess.

What do you think?